polkadot/node/network/collator-protocol/src/validator_side/claim_queue_state.rs

```diff
@@ -0,0 +1,773 @@
+/// in rp3.
+///
+/// So if we want to claim the third slot at rp 1 we are also claiming the second at rp2 and first
+/// at rp3. To track this in a simple way we can project the claim queue onto the relay blocks like
+/// this:
+///               [A]   [B]   [A] -> this is the claim queue at rp3
+///         [B]   [A]   [B]       -> this is the claim queue at rp2
+///   [A]   [B]   [A]	          -> this is the claim queue at rp1
+/// [RP 1][RP 2][RP 3][RP X][RP Y] -> relay blocks, RP x and RP Y are future blocks
+///
+/// Note that the claims at each column are the same so we can simplify this by just projecting a
+/// single claim over a block:
+///   [A]   [B]   [A]   [B]   [A]  -> claims effectively are the same
+/// [RP 1][RP 2][RP 3][RP X][RP Y] -> relay blocks, RP x and RP Y are future blocks
+///
+/// Basically this is how `ClaimQueueState` works. It keeps track of claims at each block by mapping
+/// claims to relay blocks.
+///
+/// How making a claim works?
+/// At each relay block we keep track how long is the claim queue. This is a 'window' where we can
+/// make a claim. So adding a claim just looks for a free spot at this window and claims it.
+///
+/// Note on adding a new leaf.
+/// When a new leaf is added we check if the first element in its claim queue matches with the
+/// projection on the first element in 'future blocks'. If yes - the new relay block inherits this
+/// claim. If not - this means that the claim queue changed for some reason so the claim can't be
+/// inherited. This should not happen under normal circumstances. But if it happens it means that we
+/// have got one claim which won't be satisfied in the worst case scenario.
+pub(crate) struct ClaimQueueState {
+	block_state: VecDeque<ClaimInfo>,
+	future_blocks: VecDeque<ClaimInfo>,
+}
+
+impl ClaimQueueState {
+	pub(crate) fn new() -> Self {
+		Self { block_state: VecDeque::new(), future_blocks: VecDeque::new() }
+	}
+
+	// Appends a new leaf
+	pub(crate) fn add_leaf(&mut self, hash: &Hash, claim_queue: &Vec<ParaId>) {
+		if self.block_state.iter().any(|s| s.hash == Some(*hash)) {
+			return
+		}
+
+		// First check if our view for the future blocks is consistent with the one in the claim
+		// queue of the new block. If not - the claim queue has changed for some reason and we need
+		// to readjust our view.
+		for (idx, expected_claim) in claim_queue.iter().enumerate() {
+			match self.future_blocks.get_mut(idx) {
+				Some(future_block) =>
+					if future_block.claim.as_ref() != Some(expected_claim) {
+						// There is an inconsistency. Update our view with the one from the claim
+						// queue. `claimed` can't be true anymore since the `ParaId` has changed.
+						future_block.claimed = false;
+						future_block.claim = Some(*expected_claim);
+					},
+				None => {
+					self.future_blocks.push_back(ClaimInfo {
+						hash: None,
+						claim: Some(*expected_claim),
+						claim_queue_len: 1,
```

Comment:
Generally speaking you are right but I think it's unneeded complexity.

For simplicity I use the same type (`ClaimInfo`) for `block_state` and `future_blocks`. Ideally they should be different and `claim_queue_len` should not exist in `future_blocks`. But since I am chaining `block_state` and `future_blocks` together I went for the same type. So my choice was either to have an `Option` or to set it to 1 (because there is one claim we know at this spot). `Option` doesn't bring much benefit besides lots of boilerplate code so I went for setting it to 1.

This is okay because we can't build on an unknown relay parent so this value won't be examined in practice.

Reply:
Added a comment about why 1.